Posts/2023/06/Measuring-Download-Time.md

```diff
@@ -0,0 +1,544 @@
+through Platform Invoke. This means, we need to use **C#** to create these function signatures in
+managed .NET. Here is what that code looks like:
+
+```csharp
+namespace Utilities
+{
+    using System;
+    using System.IO;
+    using System.Runtime.InteropServices;
+
+    public class WinHttp
+    {
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern IntPtr WinHttpOpen(
+            string pszAgentW,
+            uint dwAccessType,
+            string pszProxyW,
+            string pszProxyBypassW,
+            uint dwFlags
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern IntPtr WinHttpConnect(
+            IntPtr hSession,
+            string pswzServerName,
+            uint nServerPort,
+            uint dwReserved
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern IntPtr WinHttpOpenRequest(
+            IntPtr hConnect,
+            string pwszVerb,
+            string pwszObjectName,
+            string pwszVersion,
+            string pwszReferrer,
+            string ppwszAcceptTypes,
+            uint dwFlags
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern bool WinHttpSendRequest(
+            IntPtr hRequest,
+            string lpszHeaders,
+            uint dwHeadersLength,
+            IntPtr lpOptional,
+            uint dwOptionalLength,
+            uint dwTotalLength,
+            UIntPtr dwContext
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern bool WinHttpReceiveResponse(
+            IntPtr hRequest,
+            IntPtr lpReserved
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern bool WinHttpQueryDataAvailable(
+            IntPtr hRequest,
+            out uint lpdwNumberOfBytesAvailable
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern bool WinHttpReadData(
+            IntPtr hRequest,
+            IntPtr lpBuffer,
+            uint dwNumberOfBytesToRead,
+            out uint lpdwNumberOfBytesRead
+        );
+
+        [DllImport("Winhttp.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern bool WinHttpCloseHandle(IntPtr hInternet);
+    }
+}
+```
+
+Then we can use `Add-Type` to compile, and import this type in PowerShell.
+
+```powershell
+# If the code was saved in a file.
+Add-Type -TypeDefinition (Get-Content -Path 'C:\WinHttpHelper.cs' -Raw)
+
+# If it was saved using a here-string.
+Add-Type -TypeDefinition @'
+<# Code goes gere #>
+'@
+
+# Or
+
+Add-Type -TypeDefinition $typeSignature
```

Comment:
Focus on one way to import the type since you really don't explain the alternatives. Explaining the alternatives is out of scope for this post.
```suggestion
Add-Type -TypeDefinition (Get-Content -Path 'C:\WinHttpHelper.cs' -Raw)
```